scipy/optimize/_sda.py

```diff
@@ -0,0 +1,589 @@
+                        self.xmin = np.copy(self.state.current_location)
+        # End of MarkovChain loop
+
+    def local_search(self):
+        # Decision making for performing a local search
+        # based on Markov chain results
+        # If energy has been improved or no improvement since too long,
+        # performing a local search with the best Markov chain location
+        if self.state_improved:
+            # Global energy has improved, let's see if LS improved further
+            e, x = self.ofw.local_search(self.state.xbest)
+            if e < self.state.ebest:
+                self.not_improved_idx = 0
+                self.state.ebest = e
+                self.state.xbest = np.copy(x)
+                self.state.current_energy = e
+                self.state.current_location = np.copy(x)
+                return
+        # Check probability of a need to perform a LS even if no improvment
+        # (Dual annealing principle)
+        do_ls = False
+        if self.K < 90 * len(self.state.current_location):
+            pls = np.exp(self.K * (self.state.ebest - self.state.current_energy
+                                   ) / self.temperature_step)
+            if pls >= self._rs.random_sample():
+                do_ls = True
+        # Global energy not improved, let's see what LS gives
+        # on the best Markov chain location
+        if self.not_improved_idx >= self.not_improved_max_idx:
+            do_ls = True
+        if do_ls:
+            e, x = self.ofw.local_search(self.xmin)
+            self.xmin = np.copy(x)
+            self.emin = e
+            self.not_improved_idx = 0
+            self.not_improved_max_idx = self.state.current_location.size
+            if e < self.state.ebest:
+                self.state.ebest = self.emin
+                self.state.xbest = np.copy(self.xmin)
+                self.state.current_energy = e
+                self.state.current_location = np.copy(x)
+
+
+class ObjectiveFunWrapper(object):
+    """
+    Class used to wrap around the objective function in order to apply local
+    search and default gradient computation.
+    Default local minimizer is L-BFGS-B
+    """
+    def __init__(self, bounds, func, **kwargs):
+        self.func = func
+        self.nb_fun_call = 0
+        self.kwargs = kwargs
+        self.minimizer = minimize
+        self.fun_args = None
+        lu = list(zip(*bounds))
+        self.lower = np.array(lu[0])
+        self.upper = np.array(lu[1])
+        self.ls_max_iter = self.lower.size * 6
+
+        if self.ls_max_iter < 100:
+            self.ls_max_iter = 100
+        if self.ls_max_iter > 1000:
+            self.ls_max_iter = 1000
+
+        # By default, scipy L-BFGS-B is used with a custom 3 points gradient
+        # computation
+        else:
+            self.fun_args = ()
+        if not self.kwargs or 'method' not in self.kwargs:
+            self.kwargs['method'] = 'L-BFGS-B'
```

Comment:
I think the choice of the local search method deserve some thought. 

By what I know of "L-BFGS-B", the method is very good for large scale problems, but since global optimization algorithms usually cannot handle problems with too many variable (>100 variables) very well, this may not be the best choice of local search method.

What are your thoughts on this?

Reply:
Hi @antonior92, thanks a lot for your review and congratulation for being in the SciPy core team, this is great!
Regarding the choice of using "L-BFGS-B" as default local search method, in the context of the extensive benchmarks performed (see details in the PR description) this is the one that has proved the best efficiency/reliability even for small dimension problems.

Reply:
Oh, thank you very much for including the benchmarks. Your method does have some impressive results.

This choice did not matched my first intuition on the problem, but it puts my mind at easy that you have tested different methods and choose "L-BFGS-B" based on these results.